data/campaigns/fri05.wmf/scripting/mission_thread.lua

```diff
@@ -0,0 +1,229 @@
+                     trade_interesting = false
+                     break
+                  end
+               end
+               if trade_interesting then
+                  for input,amount in pairs(proposal[1]) do
+                     port:set_wares(input, port:get_wares(input) - amount)
+                  end
+                  for output,amount in pairs(proposal[2]) do
+                     port:set_wares(output, port:get_wares(output) + amount)
+                  end
+                  sleep(3600)
+               end
+            end
+         end
+      end
+      sleep(7200)
+   end
+end
+
+function check_objective_wood()
+   local o = add_campaign_objective(obj_wood)
+   while not check_for_buildings(p1, {frisians_woodcutters_house = 3, frisians_foresters_house = 4}) do
+      sleep(4273)
+   end
+   set_objective_done(o)
+   campaign_message_box(secured_wood)
+end
+
+function check_objective_block()
+   local o = add_campaign_objective(obj_block)
+   while true do
+      sleep(4916)
+      local sites = 0
+      for x = 84, 94 do
+         for y = 114, 122 do
+            local f = map:get_field(x, y)
+            if f.immovable ~= nil and f.immovable.descr.type_name == "militarysite" and f.immovable.fields[1] == f and f.owner == p1 then
+               sites = sites + 1
+               if sites >= 3 then
+                  set_objective_done(o)
+                  campaign_message_box(secured_block)
+                  return
+               end
+            end
+         end
+      end
+   end
+end
+
+function check_objective_uplands()
+   local o = add_campaign_objective(obj_uplands)
+   while true do
+      sleep(4471)
+      for _,wh in ipairs(p1:get_buildings("frisians_warehouse")) do
+         if wh.fields[1].height > 40 then
+            set_objective_done(o)
+            campaign_message_box(secured_uplands)
+            return
+         end
+      end
+   end
+end
+
+function mission_thread()
+   include "map:scripting/init_p2.lua"
+   run(check_fish)
+   campaign_message_box(intro_1)
+   campaign_message_box(intro_2)
+   campaign_message_box(intro_3)
+   campaign_message_box(intro_4)
+
+   scroll_to_field(map.player_slots[1].starting_field)
+   include "map:scripting/init_p1.lua"
+   p1:reveal_fields(map.player_slots[2].starting_field:region(9))
+   local reveal_fields = {}
+   for x=0,121 do
+      for y=17,240 do
+         local f = map:get_field(x, y)
+         p2:reveal_fields({f})
+         if p1:sees_field(f) then
+            p1:reveal_fields({f})  -- Make currently seen fields permanently visible without hiding them first
+         else
+            table.insert(reveal_fields, f)
+         end
+      end
+   end
+   run(function()
+      reveal_randomly(p1, reveal_fields, 100)
+      reveal_fields = nil
+   end)
+   sleep(5000)
+
+   campaign_message_box(intro_5)
+   campaign_message_box(intro_6)
+   campaign_message_box(intro_7)
+
+   scroll_to_field(map.player_slots[2].starting_field)
+   sleep(10000)
```

Comment:
5 seconds would be enough to see the port and the ship. 10 secs are somewhat annoyingly long